allennlp/data/iterators/lazy_iterator.py

```diff
@@ -0,0 +1,92 @@
+from typing import List, Iterator, Dict
+import itertools
+import logging
+
+from overrides import overrides
+
+from allennlp.common import Params
+from allennlp.data.iterators.data_iterator import DataIterator
+from allennlp.data.dataset import LazyDataset
+from allennlp.data.instance import Instance
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+@DataIterator.register("lazy")
+class LazyIterator(DataIterator[LazyDataset]):
+    """
+    A very basic lazy iterator, which takes a dataset and yields fixed size batches.
+    Each batch is padded to the maximum length within that batch.
+
+    Parameters
+    ----------
+    batch_size : int, optional, (default = 32)
+        The size of each batch of instances yielded when calling the iterator.
+    instances_per_epoch: int, optional, (default = None)
+        If this parameter is specified, each epoch generated by ``LazyIterator.__call__``
+        contains at most this many instances. A "cursor" is stored for each Dataset,
+        so that each epoch picks up where the previous one left off.
+
+        If the value is ``None``, then each epoch is just a full pass through the Dataset.
+    """
+    def __init__(self, batch_size: int = 32, instances_per_epoch: int = None) -> None:
+        self._batch_size = batch_size
+        self._instances_per_epoch = instances_per_epoch
+
+        # As you might use this iterator with multiple datasets,
+        # you need to store a cursor for each one.
+        self._cursors: Dict[LazyDataset, Iterator[Instance]] = {}
+
+    @overrides
+    def get_num_batches(self, _: LazyDataset) -> int:
+        # pylint: disable=no-self-use
+        # TODO(joelgrus): Figure out the right way to handle this.
+        #   This is definitely not right, but currently this is only used for tqdm,
+        #   and the only consequence is that the tqdm for iterating over an epoch
+        #   will show as a count rather than as a progress bar.
+        return 1
+
+    def _take_instances(self, dataset: LazyDataset, max_instances: int) -> Iterator[Instance]:
```

Comment:
I like this quite a bit.  It'll make it easy to extend this to over-take instances, then use a `BucketIterator` to sort them, which I think is the right thing to do eventually.  And along those lines, I'd probably call this a `LazyBasicIterator`, in analogy to our `BasicIterator`, and in preparation for a `LazyBucketIterator` that will come next.

Is there a way to have this always return `max_instances`, though, looping around immediately when you get to the end?  E.g., I'm not sure why this has to return an `Iterator`, instead of a `List`.

Reply:
yes, that way is probably a lot simpler

Reply:
but I don't want it to return a list, because max_instances could still be more than fits in memory

Reply:
I was imagining that `_create_batches` would look something like this:

```python
num_instances_yielded = 0
while num_instances_yielded < self._instances_per_epoch:
    batch = Dataset(self._take_instances(dataset, self._batch_size))
    num_instances_yielded += self._batch_size
    yield batch
```

Then, in a `LazyBucketIterator`, I can have a parameter like `self._max_instances_in_memory`, and do something like:

```python
num_instances_yielded = 0
bucket_iterator = BucketIterator(self._batch_size, ...)
while num_instances_yielded < self._instances_per_epoch:
    instances_to_take = min(self._instances_per_epoch - num_instances_yielded,
                            self._max_instances_in_memory)
    instances = self._take_instances(dataset, instances_to_take)
    yield from bucket_iterator(Dataset(instances))
    num_instances_yielded += instances_to_take
```

I think this is quite a bit simpler than what you have.  The downside is that it doesn't really work if `self._instances_per_epoch` is `None`.  I think this is fine - it seems reasonable to require specifying `instances_per_epoch` if you're using a lazy dataset.

What do you think?

Reply:
Oh, a problem with setting `instances_per_epoch` is that you probably want these to be different for training and validation, but we don't have a good mechanism for separating them...  Hmm, I'm not sure how to get around this.

Reply:
here is a suggestion for what's probably the "right" way to address this:

(1) construct Dataset "from_params"

right now in our config file, we don't really have a notion of a dataset, we just have a `train_data_path` and `test_data_path`, which we pass into `DatasetReader.read()`

probably, though, DatasetReader.read() should take a `Params` object that contains the file path plus other things like "instances_per_epoch" and "max_instances_in_memory" and constructs the Dataset accordingly. 

we could get backward compatibility with config files by checking first for a `train_data` key and then failing back to using `train_data_path` as we do currently. But we would break custom DatasetReaders, unless we called this method something other than `read`

(2) the iterator constructor / from_params takes a Dataset, and then we pass into `train()` the train_iterator and validation_iterator, rather than iterator + train_data + validation_data

this allows us to have different iteration behavior on different datasets, but still only have to specify the iterator params once (the different behaviors are driven by different dataset configs)

This (at least the second part) is sort of the path I was originally going doing with this, before I decided to change the API as little as possible. If we want to do this, I think I'd rather do it in a second PR though.

Reply:
Yeah, I think some way of sending different parameters for train and validation is necessary here.  I'm not totally sure that what you're suggesting is right (e.g., I'd probably prefer to just have two different iterator configs, instead of having the `Dataset` know about epochs and such), but it's at least close.  And you're right that this PR is great as it is, and a good first step toward what we need, and what we're talking about here goes in a separate PR.